CTF101 crypto writeups (#14)
stellarvector/blog: _writeups/2023/ctf101/substitution-ciphers.md

---

> layout: writeup

Reviewer: ```suggestion
```
`mathjax` is based on `writeup` (otherwise the layouts would be broken)

parent: CTF101
grand_parent: 2023
author: r98inver
category: crypto
tags: single-key-xor vigenere
title: "Substitution Ciphers"
subtitle: ""
write_date: 2023-10-20
last_edit_date:
layout: mathjax # Uncomment this line to enable MathJax
---

> The second crypto session of **CTF101** focused on symmetric enctryption, and more specifically substitution ciphers.

Reviewer: ```suggestion
The second crypto session of **CTF101** focused on symmetric encryption, and more specifically substitution ciphers.
```

## stellar-caesar

The first challenge was a simple Caesar Cipher, as hinted by the name. The ciphertext is `vy{vkliwlqj_olnh_wkh_dqflhqw_urpdqv}` and just putting this in CyberChef with [ROT13 shifted by -3](https://gchq.github.io/CyberChef/#recipe=ROT13(true,true,false,-3)&input=dnl7dmtsaXdscWpfb2xuaF93a2hfZHFmbGhxd191cnBkcXZ9Cg) returns the flag: `sv{shifting_like_the_ancient_romans}`. The shift of -3 could be either computed by knowing that the flag begins with `sv`, or guessed, since it was the one supposedly used by Caesar, or even bruteforced easily.

## trial-and-error

> The second challenge was the python version of the Caesar Cipher: a single byte XOR. The idea here is to simply bruteforce all the possible keys, since there are only 256 possibilities to chose one byte, and then find a way to detect when the output is correct (checking by hand 256 possibilities is still duable, but automating it is much more helpful for the next challenge and way less boring). First of all, we need a function to xor a key against our plaintext. We can use the `xor` function from `pwntools` or write a simple helper function like this:

Reviewer: ```suggestion
The second challenge was the python version of the Caesar Cipher: a single byte XOR. The idea here is to simply bruteforce all the possible keys, since there are only 256 possibilities to chose one byte, and then find a way to detect when the output is correct (checking by hand 256 possibilities is still doable, but automating it is much more helpful for the next challenge and way less boring). First of all, we need a function to xor a key against our plaintext. We can use the `xor` function from `pwntools` or write a simple helper function like this:
```

```python
def single_key_xor(data, k):
	out = bytes([x ^ k for x in data])
	return out
```

Here `k` is a number, which will simplify the next step. The simple way to check for correct output is to look for the presence of `b'sv{'` in the output.

```python
with open('out.xor', 'rb') as fh:
	data = fh.read()

for k in range(256):
	guess = single_key_xor(data, k)
	if b'sv{' in guess:
		print(guess)
		print(f'{k = }')
```

> In the decrypted text, with `k=105`, we can find the flag: `sv{t1m3s_4r3_ch4ng1ng}`. Other strategies include checking that the most frequent characters in the output are either `e` or ` `, the two most common characters in written english, or checking that the whole output consists in ASCII characters. This is an example of the former, while we will see the latter in the next challenge:

Reviewer: ```suggestion
In the decrypted text, with `k=105`, we can find the flag: `sv{t1m3s_4r3_ch4ng1ng}`. Other strategies include checking that the most frequent characters in the output are either `e` or ` `, the two most common characters in written English, or checking that the whole output consists in ASCII characters. This is an example of the former, while we will see the latter in the next challenge:
```

```python
for k in range(256):
	guess = single_key_xor(data, k)

	e_count = guess.count(b'e')
	space_count = guess.count(b' ')

	max_count = max([ guess.count(bytes([i])) for i in range(256) ])

	if e_count == max_count or space_count == max_count:
		print(f'{k = }')
		print(f'{guess = }')
```

## pygenere

Once again the name explains us a lot of the challenge: we face a *Vigenere Cipher* in its python version, i.e. a multi-key xor. Unlike the previous exercise, we cannot bruteforce the keys: the number of possible keys of length $$s$$ is $$256^s$$, so we cannot hope to go much further than maybe 3 characters. We need to come up with a smarter method. There are three main things to observe:

> - if we know the key length, we do not have to bruteforce all the possible keys, but we can build the key character by character; let say the length is three: then the first byte of the plaintext is xored with the first byte of the key, the second with the second, the third with the third, but then the fourth with the first and so on. So if we split the ciphertext in three groups, each one that got xored against a different byte of the key, we have three parallel instances of single key xor that we can target. This drasically reduce the bruteforce cost from $$256^3$$ to $$3 \times 256$$, which is very feasible;

Reviewer: ```suggestion
- if we know the key length, we do not have to bruteforce all the possible keys, but we can build the key character by character; let say the length is three: then the first byte of the plaintext is xored with the first byte of the key, the second with the second, the third with the third, but then the fourth with the first and so on. So if we split the ciphertext in three groups, each one that got xored against a different byte of the key, we have three parallel instances of single key xor that we can target. This drastically reduces the bruteforce cost from $$256^3$$ to $$3 \times 256$$, which is very feasible;
```
